logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,204 @@
+import org.zalando.logbook.HttpResponse;
+import org.zalando.logbook.Origin;
+import org.zalando.logbook.Precorrelation;
 
 /**
- * A custom {@link HttpLogFormatter} that produces JSON objects. It can be augmented with composition:
- *
- * <pre>
- * {@code
- *
- * public class CustomsFormatter implements HttpLogFormatter {
- *
- *     private final JsonHttpLogFormatter delegate;
- *
- *     public CustomsFormatter(final ObjectMapper mapper) {
- *         this.delegate = new JsonHttpLogFormatter(mapper);
- *     }
- *
- *     public String format(final Precorrelation<HttpRequest> precorrelation) throws IOException {
- *         Map<String, Object> request = delegate.prepare(precorrelation);
- *         // modify request here
- *         return delegate.format(request);
- *     }
- *
- *     public String format(final Correlation<HttpRequest, HttpResponse> correlation) throws IOException {
- *         Map<String, Object> response = delegate.prepare(correlation);
- *         // modify response here
- *         return delegate.format(response);
- *      }
- *
- * }
- * }
- * </pre>
+ * A custom {@link HttpLogFormatter} that produces JSON objects. 
  */
 @API(status = STABLE)
-public final class JsonHttpLogFormatter implements StructuredHttpLogFormatter {
+public final class JsonHttpLogFormatter implements HttpLogFormatter {
+
+    private static final String[] REPLACEMENT_CHARS;
+    
+    static {
+            REPLACEMENT_CHARS = new String[128];
+            for (int i = 0; i <= 0x1f; i++) {
+                REPLACEMENT_CHARS[i] = String.format("\\u%04x", (int) i);
+            }
+            REPLACEMENT_CHARS['"'] = "\\\"";
+            REPLACEMENT_CHARS['\\'] = "\\\\";
+            REPLACEMENT_CHARS['\t'] = "\\t";
+            REPLACEMENT_CHARS['\b'] = "\\b";
+            REPLACEMENT_CHARS['\r'] = "\\r";
+            REPLACEMENT_CHARS['\f'] = "\\f";
+      }
+    
+    @Override
+    public String format(Precorrelation precorrelation, HttpRequest request) throws IOException {
+        final String correlationId = precorrelation.getId();
+        
+        String body = request.getBodyAsString();
+
+        StringBuilder builder = new StringBuilder(body.length() + 2048);
+
+        builder.append("{\"origin\":\"");
+        if(request.getOrigin() == Origin.LOCAL) {
+            builder.append("local");
+        } else {
+            builder.append("remote");
+        }
+        builder.append("\",\"type\":\"request\",\"correlation\":\"");
+        builder.append(correlationId); 
+        builder.append("\",\"protocol\":\"");
+        encode(request.getProtocolVersion(), builder);
+        builder.append("\",\"remote\":\"");
+        encode(request.getRemote(), builder);
+        builder.append("\",\"method\":\"");
+        encode(request.getMethod(), builder);
+        builder.append("\",\"uri\":\"");
+        reconstruct(request, builder);
+        builder.append('"');
 
-    private final ObjectMapper mapper;
```

Comment:
Instead of going from `ObjectMapper` (relatively high-level) to `StringBuilder` (lowest level) maybe we could try to re-implement this using `JsonGenerator`: https://github.com/FasterXML/jackson-docs/wiki/JacksonStreamingApi#creating-generators

Extending the data structure can be achieved by allowing to register callbacks to be invoked with the generator before closing the outer most object.

Reply:
I tried that and was not too impressed with the improvement, but I guess I can dig up the code and see if it can be tweaked further.

Reply:
See https://github.com/zalando/logbook/pull/541, it is better than the original, but overall a bit disappointing result.